### ISSUE TITLE
Optional download of nodewebkit

### ISSUE THREAD

**Issue by JimmyBoh** [intelligentgolf/karma-nodewebkit-launcher#24]
This launcher is out of date, especially with nodewebkit turning in to `nw.js`. Instead of including a dependency on `nodwebkit`, instead require `NODEWEBKIT_BIN` or optionally download the binaries. That way you don't have to maintain the version updates (except for the optional download functionality). The other benefit is in continuous integration, the downloading is the slowest part of my build.

**Comment by michalc** [comment]
Hi Jimmy,

Optional sounds good. What would the `package.json` look like to make the download of the binaries optional? For a user of this launcher, how would they then tell npm to download (or not) the binaries?

Thanks

**Comment by JimmyBoh** [comment]
I would think that we would remove the dependency on `nodewebkit` and use options passed in to the launcher to either point to the executable or download it. I wonder if calling `npm install nodewebkit` from within the launcher would be a good approach?

**Comment by megawac** [comment]
Why not just put

```json
peerDependencies: {
   "nodewebkit": "*"
}
```

**Comment by michalc** [comment]
@JimmyBoh Will have a look into the installing nodewebkit from within the launcher. I see at http://stackoverflow.com/a/15957574/1319998 you can use npm programatically, so no need to invoke it via shell commands. I have to say though, it makes me a bit nervous to have such a dependency "hidden" within config, and then if downloading is part of launching the browser from the point of view of karma, launching timeouts might need to be extended, which then makes it difficult to find out if there is some problem _really_ launching the browser. The launcher for phantomjs mentions its dependency on phantomjs explicity https://github.com/karma-runner/karma-phantomjs-launcher/blob/master/package.json for example.

@megawac Would specifying it as a peerDependency make it optional?

**Comment by michalc** [comment]
@JimmyBoh Regarding nodewebkit turning into nw.js, I see nw.js is still alpha, so I don't think it will be a good idea to bump to that version quite yet. How do you feel about for now it being changed to:

    "nodewebkit": "^0.11.0"

Regarding the download taking a lot of time... this might be me slightly passing the buck, but I wonder if it should be up to https://github.com/nwjs/npm-installer/ to somehow cache the downloaded binary, or even up to npm itself? I can't help feeling it shouldn't be up to a plugin such as this to handle anything regarding installing its dependencies.

**Comment by michalc** [comment]
@JimmyBoh I've submitted a pull request to the nwjs installer at https://github.com/nwjs/npm-installer/pull/7 that allows a custom binary source location. I wonder if you can use this to maintain some sort of local mirror of the binaries?

If you want to use it right now, in the `package.json` in the project you're actually testing, have a dependency to my specific branch:

    "karma-nodewebkit-launcher": "michalc/karma-nodewebkit-launcher#custom-url-base",

And then you can specify a custom location for the binaries, for example, in the local project `.npmrc` file:

    nwjs_urlbase=http://my.own.location/somewhere

Or you can do this via the user's `.npmrc` file, or an evironment variable. See the pull request for details

**Comment by JimmyBoh** [comment]
Thanks @michalc, that definitely seems like a nice feature. Ideally I'd download the current binary and check it in to source control. Then I would point `karma-nodewebkit-launcher` at the nw executable to skip downloading.  Can the URL base point to a local folder with the files? Peer dependencies might be the way to go, but I haven't tested it to see if it tries to download it again.

Honestly it's been a while since I looked in to the issue, but I know that [this fork](https://github.com/nDmitry/karma-nodewebkit-launcher) seemed to work for me (latest version of nw.js). I hate to see so many forks of the `karma-nodewebkit-launcher` but no pull requests to unify them.

**Comment by iamdenny** [comment]
Please remove "nodewebkit": "^0.8.5" in package.json.
It takes so long to install. and the newest one is already installed.
thanks.

**Comment by michalc** [comment]
@iamdenny and @JimmyBoh 

I've created a branch at https://github.com/michalc/karma-nodewebkit-launcher/tree/file-url that avoids the download by accepting an option `nwjs_urlbase` that tells it where the zip / .tar.gz file that would be downloaded is located: you can now give it a `file://` URL, as well as a `http://` URL.

To use it, you can replace the `karma-nodewebkit-launcher` line in your package.json with

    "karma-nodewebkit-launcher": "michalc/karma-nodewebkit-launcher#file-url"

and see the README at https://github.com/michalc/karma-nodewebkit-launcher/tree/file-url for info on how to set the option. This branch doesn't really do anything different in the launcher code, other than depend on what I've submitted to https://github.com/nwjs/npm-installer/pull/11 , which processes the option. If/when the pull request is submitted, I can merge the branch and bump the version here / in npm.

At the moment I'm still not very pro removing the nodewebkit dependency on the installer, in order to preserve the behaviour that it should just work out of the box (albeit slowly). I'm much more pro ways users can then optionally optimise it. 

Regarding the peerDependency, I've had a go, but so far it doesn't seem to stop it from downloading the binary.

**Comment by JimmyBoh** [comment]
@michalc Sounds good, hopefully the pull request doesn't sit forever!

**Comment by michalc** [comment]
@JimmyBoh The pull request was merged, and the nw installer version bumped to use it (to 0.12.0-1). I've merged the branch from https://github.com/michalc/karma-nodewebkit-launcher/tree/file-url, and bumped the version in NPM (to 0.0.12)

So v0.0.12 (or later) of karma-nodewebkit-installer should be able to grab the binary from file:// or http:// location. See the readme for instructions.

For future readers, the "12" in common with the current installer version 0.0.12 and the nw installer 0.12-1 is a coincidence :-)
